pkg/backend/filestate/store.go

```diff
@@ -232,49 +235,63 @@ func (p *projectReferenceStore) ListProjects() ([]tokens.Name, error) {
 func (p *projectReferenceStore) ListReferences() ([]*localBackendReference, error) {
 	// The first level of the bucket is the project name.
 	// The second level of the bucket is the stack name.
-	path := StacksDir
-
-	projects, err := p.ListProjects()
-	if err != nil {
-		return nil, err
-	}
+	ctx := context.TODO()
+	prefix := StacksDir + "/"
+	iter := p.bucket.List(&blob.ListOptions{
+		Prefix: prefix,
+		// Don't set the Delimiter.
+		// This will treat the entire bucket as a flat list,
+		// returning only files under the prefix.
+	})
 
 	var stacks []*localBackendReference
-	for _, projName := range projects {
-		// TODO: Could we improve the efficiency here by firstly making listBucket return an enumerator not
-		// eagerly collecting all keys into a slice, and secondly by getting listBucket to return all
-		// descendent items not just the immediate children. We could then do the necessary splitting by
-		// file paths here to work out project names.
-		projectFiles, err := listBucket(p.bucket, filepath.Join(path, projName.String()))
+	for {
+		file, err := iter.Next(ctx)
```

Comment:
Is this fast on remote buckets?

Reply:
Should be, means gocloud doesn't _have_ to load the whole bucket at once.

Reply:
It'll use whatever the underlying pagination is.
The driver.Bucket interface only has a [ListPaged method](https://pkg.go.dev/gocloud.dev@v0.29.0/blob/driver#Bucket.ListPaged), so the way this'll work is S3 will give you the first, say, 100 items, then Next will iterate through them in memory until it runs out and needs the next 100.

And the "flatten" behavior is the bucket's default: S3 doesn't have a concept of directories beyond buckets, so "foo/bar" is just a key; "/" has no meaning to S3 if you don't set the delimiter.